File: src/main/services/dns.ts
import { promises as dns } from "node:dns";
import { getCachedDns, setCachedDns } from "./persistence";
import { getLogger } from "./logger";

const log = getLogger();

export interface ReverseDnsOptions {
  forceRefresh?: boolean;
}

export async function resolveReverseDns(
  ipAddress: string,
  options?: ReverseDnsOptions
): Promise<string | undefined> {
  if (!ipAddress) {
    return undefined;
  }

  if (!options?.forceRefresh) {
    const cached = getCachedDns(ipAddress);
    if (cached) {
      return cached;
    }
  }

  try {
    const [host] = await dns.reverse(ipAddress);
    if (host) {
      if (!options?.forceRefresh) {
        setCachedDns(ipAddress, host);
      }
      return host;
    }
  } catch (error) {
    log.debug("Reverse DNS lookup failed", ipAddress, error);
  }

  return undefined;
}

File: src-tauri/src/commands/dns.rs
use crate::commands::persistence::SharedStore;

/// Perform a reverse DNS lookup for an IP address, with caching.
pub async fn resolve_reverse_dns(
    ip: &str,
    force_refresh: bool,
    store: &SharedStore,
) -> Option<String> {
    if ip.is_empty() {
        return None;
    }

    if !force_refresh {
        let guard = store.lock().unwrap();
        if let Some(cached) = guard.get_cached_dns(ip) {
            return Some(cached);
        }
    }

    // Use dns-lookup for reverse DNS
    match dns_lookup::lookup_addr(&ip.parse().ok()?) {
        Ok(host) => {
            if !host.is_empty() && host != ip {
                if !force_refresh {
                    let mut guard = store.lock().unwrap();
                    guard.set_cached_dns(ip, &host);
                }
                Some(host)
            } else {
                None
            }
        }
        Err(e) => {
            log::debug!("Reverse DNS lookup failed for {}: {}", ip, e);
            None
        }
    }
}
